ksqldb-engine/src/main/java/io/confluent/ksql/internal/UtilizationMetricsListener.java

```diff
@@ -0,0 +1,193 @@
+    private final String CONSUMER_METRICS = "consumer-metrics";
+    private final String PRODUCER_METRICS = "producer-metrics";
+    private final String THREAD_ID = "thread-id";
+
+    private final List<KafkaStreams> kafkaStreams;
+    private final Logger logger = LoggerFactory.getLogger(UtilizationMetricsListener.class);
+    private final List<String> metrics;
+    private final Time time;
+    private long lastSampleTime;
+
+    private final Map<String, Double> previousFlushTime;
+    private final Map<String, Double> previousConsumerIOTime;
+    private final Map<String, Double> previousProducerBufferWaitTime;
+
+    public UtilizationMetricsListener(){
+        this.kafkaStreams = new ArrayList<>();
+        this.metrics = new LinkedList<>();
+        initializeMetrics();
+        time = Time.SYSTEM;
+        lastSampleTime = time.milliseconds();
+        previousFlushTime = new HashMap<>();
+        previousConsumerIOTime = new HashMap<>();
+        previousProducerBufferWaitTime = new HashMap<>();
+    }
+
+    // for testing
+    public UtilizationMetricsListener(final List<KafkaStreams> streams, final Time time, final long lastSample) {
+        this.kafkaStreams = streams;
+        this.metrics = new LinkedList<>();
+        initializeMetrics();
+        this.time = time;
+        lastSampleTime = lastSample;
+        previousFlushTime = new HashMap<>();
+        previousConsumerIOTime = new HashMap<>();
+        previousProducerBufferWaitTime = new HashMap<>();
+    }
+
+    @Override
+    public void onCreate(
+            final ServiceContext serviceContext,
+            final MetaStore metaStore,
+            final QueryMetadata queryMetadata) {
+        kafkaStreams.add(queryMetadata.getKafkaStreams());
+    }
+
+    @Override
+    public void onDeregister(final QueryMetadata query) {
+        final KafkaStreams streams = query.getKafkaStreams();
+        kafkaStreams.remove(streams);
+        for (ThreadMetadata thread : streams.metadataForLocalThreads()) {
+            final String name = thread.threadName();
+            previousFlushTime.remove(name);
+            previousConsumerIOTime.remove(name);
+            previousProducerBufferWaitTime.remove(name);
+        }
+    }
+
+    @Override
+    public void run() {
+        logger.info("Reporting CSU thread level metrics");
+        processingRatio();
+
+    }
+
+    // public for testing
+    public double processingRatio() {
+        long sampleTime = time.milliseconds();
+
+        final long windowSize = sampleTime - lastSampleTime;
+        final long windowStart = lastSampleTime;
+        double blockedTime;
+        blockedTime = sampleTime - lastSampleTime;
+        for (KafkaStreams stream : kafkaStreams) {
+            for (ThreadMetadata thread : stream.metadataForLocalThreads()) {
+                final double threadTimes = getProcessingRatio(thread.threadName(), stream, windowStart, windowSize);
```

Comment:
nit: `threadTimes` here is a bit confusing. Should it just be `threadBlockedTime`?

Also I think it's cleaner to initialize `blockedTime` as `windowSize`, and then `blockedTime = Math.min(threadBlockedTime, blockedTime);` below.

Furthermore I'm wondering why we want to just use the min among all threads within instances to indicate the node's processing ratio? What about if some instances have very low processing ratio but others on the same pod have reasonably high processing ratio?